x-pack/filebeat/input/s3/collector.go

```diff
@@ -0,0 +1,635 @@
+)
+
+type s3Collector struct {
+	cancellation context.Context
+	logger       *logp.Logger
+
+	config            *config
+	visibilityTimeout int64
+
+	sqs       sqsiface.ClientAPI
+	s3        s3iface.ClientAPI
+	publisher beat.Client
+}
+
+type s3Info struct {
+	name                     string
+	key                      string
+	region                   string
+	arn                      string
+	expandEventListFromField string
+}
+
+type bucket struct {
+	Name string `json:"name"`
+	Arn  string `json:"arn"`
+}
+
+type object struct {
+	Key string `json:"key"`
+}
+
+type s3BucketObject struct {
+	bucket `json:"bucket"`
+	object `json:"object"`
+}
+
+type sqsMessage struct {
+	Records []struct {
+		EventSource string         `json:"eventSource"`
+		AwsRegion   string         `json:"awsRegion"`
+		EventName   string         `json:"eventName"`
+		S3          s3BucketObject `json:"s3"`
+	} `json:"Records"`
+}
+
+type s3Context struct {
+	mux  sync.Mutex
+	refs int
+	err  error // first error witnessed or multi error
+	errC chan error
+}
+
+var (
+	// The maximum number of messages to return. Amazon SQS never returns more messages
+	// than this value (however, fewer messages might be returned).
+	maxNumberOfMessage int64 = 10
+
+	// The duration (in seconds) for which the call waits for a message to arrive
+	// in the queue before returning. If a message is available, the call returns
+	// sooner than WaitTimeSeconds. If no messages are available and the wait time
+	// expires, the call returns successfully with an empty list of messages.
+	waitTimeSecond int64 = 10
+)
+
+func (c *s3Collector) run() {
+	defer c.logger.Infof("s3 input worker for '%v' has stopped.", c.config.QueueURL)
+	c.logger.Infof("s3 input worker has started. with queueURL: %v", c.config.QueueURL)
+	fmt.Println("c.cancellation.Err() = ", c.cancellation.Err())
+	for c.cancellation.Err() == nil {
+		// receive messages from sqs
+		output, err := c.receiveMessage(c.sqs, c.visibilityTimeout)
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == awssdk.ErrCodeRequestCanceled {
+				continue
+			}
+			c.logger.Error("SQS ReceiveMessageRequest failed: ", err)
+			time.Sleep(time.Duration(waitTimeSecond) * time.Second)
+			continue
+		}
+
+		fmt.Println("output= ", output)
+		if output == nil || len(output.Messages) == 0 {
+			c.logger.Debug("no message received from SQS:", c.config.QueueURL)
+			continue
+		}
+
+		// process messages received from sqs, get logs from s3 and create events
+		c.processor(c.config.QueueURL, output.Messages, c.visibilityTimeout, c.s3, c.sqs)
+	}
+}
+
+func (c *s3Collector) processor(queueURL string, messages []sqs.Message, visibilityTimeout int64, svcS3 s3iface.ClientAPI, svcSQS sqsiface.ClientAPI) {
+	var wg sync.WaitGroup
+	numMessages := len(messages)
+	c.logger.Debugf("Processing %v messages", numMessages)
+	wg.Add(numMessages * 2)
+
+	// process messages received from sqs
+	for i := range messages {
+		errC := make(chan error)
```

Comment:
Instead of using a bare `WaitGroup` you could use `MultiErrGroup` from `github.com/elastic/go-concert`: https://github.com/elastic/go-concert/blob/master/unison/multierrgroup.go#L30

The code could get smaller and we can avoid the issues that might come from forgetting to call `Done`, etc.

Reply:
Done! Thank you for the suggestion!